cocotb/decorators.py

```diff
@@ -460,18 +461,12 @@ def __init__(self, f, timeout_time=None, timeout_unit=None,
         type(self)._id_count += 1
 
         if timeout_time is not None:
-            co = coroutine(f)
+            co = types.coroutine(f)
 
             @functools.wraps(f)
-            def f(*args, **kwargs):
+            async def f(*args, **kwargs):
                 running_co = co(*args, **kwargs)
-                try:
-                    res = yield cocotb.triggers.with_timeout(running_co, self.timeout_time, self.timeout_unit)
-                except cocotb.result.SimTimeoutError:
-                    running_co.kill()
-                    raise
-                else:
-                    return res
+                return await cocotb.triggers.with_timeout(running_co, self.timeout_time, self.timeout_unit)
```

Comment:
This change was possible because `running_co` never touches the scheduler until it is `fork`ed by `First._wait`, which cleans up after the first trigger fires. This change was necessary to get the above `types.coroutine` line to work.

Reply:
I'm a bit confused by the fact that you no longer kill the coroutine here.

Reply:
It's no longer necessary thanks to the improved semantics offered by the `with_timeout` rewrite. `running_co` doesn't get scheduled, and doesn't need to be killed.

Reply:
Here's the real answer - you changed the semantics of `First` in this patch to kill non-first coroutines. I don't think we want that change.